File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiKnob.cpp
#include "StdAfx.h"
#include "UiKnob.h"

#define ThisMethod($x) &UiKnob::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiKnob::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiKnob> c_obj;
	}

	void UiKnob::DefineControl()
	{
		AutoAddMethod( SetText );
		AutoAddMethod( GetText, WrapObjectGeneric );

		AutoAddMethod( ValueSetRange );
		AutoAddMethod( ValueGetRange, WrapObjectGeneric );

		AutoAddMethod( ValueSetDefault );
		AutoAddMethod( ValueGetDefault, WrapObjectGeneric );

		AutoAddMethod( ValueSet );
		AutoAddMethod( ValueGet, WrapObjectGeneric );

		AutoAddMethod( StepSetUnlimited );
		AutoAddMethod( StepGetUnlimited, WrapObjectGeneric );

		AutoAddMethod( ValueSetCycle );
		AutoAddMethod( ValueGetCycle, WrapObjectGeneric );

		AutoAddMethod( StepSetLarge );
		AutoAddMethod( StepGetLarge, WrapObjectGeneric );

		AutoAddMethod( StepSetSmall );
		AutoAddMethod( StepGetSmall, WrapObjectGeneric );

		AutoAddMethod( ValueSetSplit );
		AutoAddMethod( ValueGetSplit, WrapObjectGeneric );

		AutoAddMethod( ValueSetTick );
		AutoAddMethod( ValueGetTick, WrapObjectGeneric );

		AutoAddMethod( RingSetColor );
		AutoAddMethod( RingGetColor, WrapObjectGeneric );

		AutoAddMethod( RingSetDefaultColor );
		AutoAddMethod( RingGetDefaultColor, WrapObjectGeneric );

		AutoAddMethod( RingSetTickColor );
		AutoAddMethod( RingGetTickColor, WrapObjectGeneric );

		AutoAddMethod( RingSetVisual );
		AutoAddMethod( RingGetVisual, WrapObjectGeneric );

		AutoAddMethod( OnChange, WrapObjectGeneric );
		AutoAddMethod( OnRotateBegin, WrapObjectGeneric );
		AutoAddMethod( OnRotateEnd, WrapObjectGeneric );
	}

	U1 UiKnob::Ctor( UiWindow * p, Napi::Value v )
	{
		if ( !__CreateControl( p, v ) )
			return false;

		return true;
	}

	void UiKnob::__OnChange( Ui::IKnob & sender )
	{
		m_OnChange( this );
	}

	void UiKnob::__OnRotateBegin( Ui::IKnob & sender )
	{
		m_OnRotateBegin( this );
	}

	void UiKnob::__OnRotateEnd( Ui::IKnob & sender, U1 bCanceled )
	{
		m_OnRotateEnd( this, bCanceled );
	}

	UiKnob* UiKnob::OnChange      /**/( OnGeneric_t   /**/ && fn ) { SetEventCallback<Ui::IKnob::OnChange      /**/>( m_OnChange      /**/, std::move( fn ), MakeThisFunc( __OnChange      /**/ ) ); return this; }
	UiKnob* UiKnob::OnRotateBegin /**/( OnGeneric_t   /**/ && fn ) { SetEventCallback<Ui::IKnob::OnRotateBegin /**/>( m_OnRotateBegin /**/, std::move( fn ), MakeThisFunc( __OnRotateBegin /**/ ) ); return this; }
	UiKnob* UiKnob::OnRotateEnd   /**/( OnRotateEnd_t /**/ && fn ) { SetEventCallback<Ui::IKnob::OnRotateEnd   /**/>( m_OnRotateEnd   /**/, std::move( fn ), MakeThisFunc( __OnRotateEnd   /**/ ) ); return this; }

}

File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiKnob.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "./UiWindow";
import { Vec2, Vec4 } from "../../Math/Vector";
import { IVisual } from "../UiVisual";

export interface IKnob extends IControl {
    new (window: Window, key?: string | number): IKnob;

    SetText(s: string): Knob;
    GetText(): string;

    ValueSetRange(v: Vec2): Knob;
    ValueGetRange(): Vec2;

    ValueSetDefault(f: number): Knob;
    ValueGetDefault(): number;

    ValueSet(f: number): Knob;
    ValueGet(): number;

    StepSetUnlimited(f: number): Knob;
    StepGetUnlimited(): number;

    ValueSetCycle(b: boolean): Knob;
    ValueGetCycle(): boolean;

    StepSetLarge(f: number): Knob;
    StepGetLarge(): number;

    StepSetSmall(f: number): Knob;
    StepGetSmall(): number;

    ValueSetSplit(f: number): Knob;
    ValueGetSplit(): number;

    ValueSetTick(t: number[]): Knob;
    ValueGetTick(): number[];

    RingSetColor(v: Vec4): Knob;
    RingGetColor(): Vec4;

    RingSetDefaultColor(v: Vec4): Knob;
    RingGetDefaultColor(): Vec4;

    RingSetTickColor(v: Vec4): Knob;
    RingGetTickColor(): Vec4;

    RingSetVisual(v: IVisual): IVisual;
    RingGetVisual(): IVisual;

    OnChange(fn: (sender: Knob) => void): Knob;
    OnRotateBegin(fn: (sender: Knob) => void): Knob;
    OnRotateEnd(fn: (sender: Knob, bCanceled: boolean) => void): Knob;
}

export class Knob extends (AveLib.UiKnob as IKnob) {
    private m_RingVisual: IVisual;

    RingSetVisual(v: IVisual) {
        this.m_RingVisual = v;
        return super.RingSetVisual(v);
    }

    ValueGetRange(): Vec2 {
        return Vec2.FromNative(super.ValueGetRange());
    }

    RingGetColor(): Vec4 {
        return Vec4.FromNative(super.RingGetColor());
    }

    RingGetDefaultColor(): Vec4 {
        return Vec4.FromNative(super.RingGetDefaultColor());
    }

    RingGetTickColor(): Vec4 {
        return Vec4.FromNative(super.RingGetTickColor());
    }
}
